### PR TITLE
Challenge 13 - Use `Action Dispatch` to route requests

### PR DESCRIPTION
**Pull request by adrianna-chang-shopify**
In this challenge, we introduce the Action Dispatch framework to handle parsing our incoming HTTP requests and routing them to the correct endpoint.

But what is an endpoint?!
- Conforms to the same specification as a Rack application, in that it expects to be called with an `environment` hash and then return a three-element array
- We initialize an [`ActionDispatch::Routing::RouteSet`](https://www.rubydoc.info/docs/rails/ActionDispatch/Routing/RouteSet), which can draw routes using a `#draw` method that takes a block declaring the routes to draw
- The[ DSL for mapping routes ](https://www.rubydoc.info/docs/rails/ActionDispatch/Routing/Mapper/HttpHelpers) uses `<action> <path>, to: <endpoint>`, where `action` is `get`, `put`, `post`, etc...

The `RouteSet` is super duper smart and can even route based on wildcard patterns:
```ruby
      match '*path', via: :all, to: -> environment {
        response = Rack::Response.new
        response.write "Sorry, I don’t know what #{environment['PATH_INFO']} is"
        response.content_type = 'text/plain'
        response.status = 404
        response.finish
      }
```

**BIG TAKEAWAY!**
> The end result is to stitch several small Rack applications (“endpoints”) together into a single large Rack application which can serve responses by directing requests to the correct endpoint.

The router essentially stitches together these mini Rack apps (endpoints), and then provides a `.call` method that takes the `environment`, and defers the work to the appropriate endpoint.

We can wrap the call to the router inside _another_ Rack application like so: 
```ruby
app = lambda { |environment| router.call(environment) }
```
This allows us to decorate the app with additional behaviour by wrapping it in middleware-esque classes! See https://github.com/adrianna-chang-shopify/learning-project/commit/f2dbf9921a1f214b50d4bc0dc0d145ee475bc79e for all the fun Tom and I had implementing these 🎉 

**config.ru fun 🦘**
https://github.com/adrianna-chang-shopify/learning-project/commit/34e2322dbec017fac6d9ef8245aa931f21c21c11 refactors our main app file to actually have an invokable endpoint that we can call from config.ru. @tomstuart let me know if this is an okay way to do this! To sum, we:
- Make our rack application an actual Ruby object in the `action_dispatch.rb` file - on initialize, it creates a routeset and draws the routes we want
- It implements `call`, which takes an environment argument and returns the results of calling our router (conforms to the Rack specification)
- In `config.ru`, we require our app file, implement our middleware logic, and then use `run MyApp.new`

Now, we can just use `puma` from the console and it all works! 🎉  (I couldn't figure out how to specify the port number / verbose mode when using a config.ru 🤔 )

### CODE REVIEW

File: challenge_13/action_dispatch.rb [hunk 1]
```diff
@@ -43,31 +50,28 @@ class Blog < ActiveRecord::Base; end
...
-    response.content_type = 'text/html'
+    Blog.create!(request.params)
+    response.redirect('/show-data', 303)
     response.finish
-  elsif request.get?
-    response.write "method is #{request.request_method}, path is #{request.path}"
+  }
+  match '*path', via: :all, to: -> environment {
+    response = Rack::Response.new
+    response.write "Sorry, I don’t know what #{environment['PATH_INFO']} is"
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: I missed that this code (copied from my doc, I think) is referencing a key in the `environment` hash directly. It’d be nicer to access the hash indirectly through `Rack::Request` for the same “less knowledge is better” reason we started doing that earlier.

File: challenge_13/action_dispatch.rb [hunk 2]
```diff
@@ -23,15 +24,21 @@ class Blog < ActiveRecord::Base; end
...
 
-app = lambda { |environment|
-  puts 'Rack app got a request!'
-  # Build Rack request
-  request = Rack::Request.new(environment)
-
-  # Build Rack response
-  response = Rack::Response.new
-
-  if request.get? && request.path == '/show-data'
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: Ultra nitpick: it was a little hard to follow this diff because the endpoints changed order, e.g. previously `/show-data` was first but now `/` is first. If rearranging these was important to you then you could’ve first made a commit reordering the code as-is and then a separate commit converting to `ActionDispatch::Routing::RouteSet` (or vice versa) rather than combining both changes. No big deal though, I got there in the end 😄

File: challenge_13/action_dispatch.rb [hunk 3]
```diff
@@ -69,10 +69,34 @@ class Blog < ActiveRecord::Base; end
...
+  end
+
+  def call(environment)
+    response = Rack::Response[*@app.call(environment)]
+    response.content_type = "#{response.content_type}; charset=#{@charset}"
+    response.finish
+  end
+end
+
+app = lambda { |environment| router.call(environment) }
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: This line gets removed in the next commit anyway, but note that it’s effectively the same as `app = router` because `lambda { |*args| whatever.call(*args) }` adds nothing and is therefore indistinguishable from just `whatever` (mathematically they are [extensionally equal](https://en.wikipedia.org/wiki/Extensionality) 🤓 ). So in practice you could’ve removed this line entirely and just done `app = CharsetMiddleware.new(router)` below, or indeed written `app = router` if you wanted to make it explicit that the router itself _is_ the application at the bottom of the middleware stack.

File: challenge_13/config.ru [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def call(environment)
+    response = Rack::Response[*@app.call(environment)]
+    response.content_type = "#{response.content_type}; charset=#{@charset}"
+    response.finish
+  end
+end
+
+use LoggerMiddleware, 'Rack app got a request!'
```
Comment (tomstuart) [inline challenge_13/config.ru:28]: lol, guess that “extra constructor parameters” thing works 😅

File: challenge_13/action_dispatch.rb [hunk 5]
```diff
@@ -69,10 +69,34 @@ class Blog < ActiveRecord::Base; end
...
+
+class CharsetMiddleware
+  def initialize(app, charset: 'UTF-8')
+    @app = app
+    @charset = charset
+  end
+
+  def call(environment)
+    response = Rack::Response[*@app.call(environment)]
+    response.content_type = "#{response.content_type}; charset=#{@charset}"
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: As a thought experiment: how could we make this middleware [liberal in what it accepts](https://en.wikipedia.org/wiki/Robustness_principle) by having it deal gracefully with situations where the response’s `Content-Type` header already contains a `charset` [media type parameter](https://tools.ietf.org/html/rfc7231#section-3.1.1.1)? ([Hint](https://github.com/rack/rack/blob/v2.2.2/lib/rack/response.rb#L195-L197).)

File: challenge_13/action_dispatch.rb [hunk 6]
```diff
@@ -69,10 +69,34 @@ class Blog < ActiveRecord::Base; end
...
+  def initialize(app, string)
+    @app = app
+    @string = string
+  end
+
+  def call(environment)
+    puts @string
+    @app.call(environment)
+  end
+end
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: If you’re curious (😂 ), check out [`Rack::CommonLogger`](https://github.com/rack/rack/blob/v2.2.2/lib/rack/common_logger.rb) which implements a more sophisticated version of the same idea. In particular I think you’ll be interested in their technique for solving the problem of “how do we write a log message when the request is completely over, given that (because of streaming responses) that might be much later than the time when `#call` actually returns?”. Note also the mechanism by which it integrates with the confusingly-named [`Rack::Logger`](https://github.com/rack/rack/blob/v2.2.2/lib/rack/logger.rb), whose job is to _provide_ a logger to subsequent middleware.

File: challenge_13/action_dispatch.rb [hunk 7]
```diff
@@ -69,10 +69,34 @@ class Blog < ActiveRecord::Base; end
...
+    @app = app
+    @charset = charset
+  end
+
+  def call(environment)
+    response = Rack::Response[*@app.call(environment)]
+    response.content_type = "#{response.content_type}; charset=#{@charset}"
+    response.finish
+  end
+end
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: Again, you have independently invented something pretty similar to [`Rack::Charset`](https://github.com/adamluzsi/rack-charset/blob/v1.0.0/lib/rack/charset.rb) here. Conventions work! 😄

File: challenge_13/action_dispatch.rb [hunk 8]
```diff
@@ -24,79 +24,61 @@ class Blog < ActiveRecord::Base; end
...
+class MyApp
+  def initialize
+    @router = ActionDispatch::Routing::RouteSet.new
+    draw_routes
   end
 
   def call(environment)
-    puts @string
-    @app.call(environment)
+    @router.call(environment)
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb:33]: What you’ve done works well, but note that there is a bit of optional boilerplate here — `MyApp#call` doesn’t do anything useful, it’s just delegating to `@router.call`. So you could just as well have said `MyApp = ActionDispatch::Routing::RouteSet.new; MyApp.draw do … end` and cut out the need to instantiate a class altogether. (Of course, implementing the app as a class gives you the advantages of being able to parameterise its constructor and make multiple instances of it in a single process, neither of which you’re doing right now but could be desirable in other situations.)

File: challenge_13/action_dispatch.rb [hunk 9]
```diff
@@ -24,79 +24,61 @@ class Blog < ActiveRecord::Base; end
...
+    end
   end
 end
-
-app = lambda { |environment| router.call(environment) }
-app = CharsetMiddleware.new(app)
-app = LoggerMiddleware.new(app, 'Rack app got a request!')
-
-# Run the Puma app server with our web application
-Rack::Handler::Puma.run(app, Port: 1234, Verbose: true)
```
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: As you noted, we’ve lost the ability to control the port number and verbosity ourselves now that we’re no longer starting Puma, but that’s kind of a good thing because instead it’s up to the person _running the application_:

```
% puma --help
puma <options> <rackup file>
    -b, --bind URI                   URI to bind to (tcp://, unix://, ssl://)
        --bind-to-activated-sockets [only]
                                     Bind to all activated sockets
    -C, --config PATH                Load PATH as a config file
        --control-url URL            The bind url to use for the control server. Use 'auto' to use temp unix server
        --control-token TOKEN        The token to use as authentication for the control server
        --debug                      Log lowlevel debugging information
        --dir DIR                    Change to DIR before starting
    -e, --environment ENVIRONMENT    The environment to run the Rack app on (default development)
    -f, --fork-worker=[REQUESTS]     Fork new workers from existing worker. Cluster mode only
                                     Auto-refork after REQUESTS (default 1000)
    -I, --include PATH               Specify $LOAD_PATH directories
    -p, --port PORT                  Define the TCP port to bind to
                                     Use -b for more advanced options
        --pidfile PATH               Use PATH as a pidfile
        --preload                    Preload the app. Cluster mode only
        --prune-bundler              Prune out the bundler env if possible
        --extra-runtime-dependencies GEM1,GEM2
                                     Defines any extra needed gems when using --prune-bundler
    -q, --quiet                      Do not log requests internally (default true)
    -v, --log-requests               Log requests as they occur
    -R, --restart-cmd CMD            The puma command to run during a hot restart
                                     Default: inferred
    -S, --state PATH                 Where to store the state details
    -t, --threads INT                min:max threads to use (default 0:16)
        --early-hints                Enable early hints support
    -V, --version                    Print the version information
    -w, --workers COUNT              Activate cluster mode: How many worker processes to create
        --tag NAME                   Additional text to display in process listing
        --redirect-stdout FILE       Redirect STDOUT to a specific file
        --redirect-stderr FILE       Redirect STDERR to a specific file
        --[no-]redirect-append       Append to redirected files
    -h, --help                       Show help
```

So a user who doesn’t care can type simply `puma` and get the defaults, but another can say `puma --port 1234 --log-requests` if they want the previous behaviour.

File: challenge_13/config.ru [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    @app = app
+    @charset = charset
+  end
+
+  def call(environment)
+    response = Rack::Response[*@app.call(environment)]
+    response.content_type = "#{response.content_type}; charset=#{@charset}"
+    response.finish
+  end
+end
```
Comment (tomstuart) [inline challenge_13/config.ru]: To state the obvious: you could easily move these pieces of middleware into their own file(s) and `require_relative` them if you wanted to keep `config.ru` small. But since this “configuration” file is literally just Ruby, it’s also fine to have as much code here as you like!

File: challenge_13/action_dispatch.rb (hunk 9 above)
Comment (tomstuart) [inline challenge_13/action_dispatch.rb]: psst, now we can remove another `require` 🔥

**Review by tomstuart — APPROVED** [review]
Another excellent and thorough solution, well done! I particularly love how the combination of `ActiveRecord::Schema.define` and `config.ru` now makes this app incredibly easy to run: just type `puma` and you’re cooking. 🎉

File: challenge_13/action_dispatch.rb (hunk 2 above)
Comment (adrianna-chang-shopify) [reply challenge_13/action_dispatch.rb]: Yes, my bad! I chose to refactor `/` first because it seemed like the simplest scenario to migrate over, and forgot that would mess up the diffs 😅

File: challenge_13/action_dispatch.rb (hunk 3 above)
Comment (adrianna-chang-shopify) [reply challenge_13/action_dispatch.rb]: Verrrry coool 🤓  Lambdas are pretty mind-bendy! Being able to just assign `app = router` as the entrypoint to the middleware stack is pretty neat 🎉

File: challenge_13/action_dispatch.rb (hunk 9 above)
Comment (adrianna-chang-shopify) [reply challenge_13/action_dispatch.rb]: Oh yay!!! 🙌

File: challenge_13/action_dispatch.rb (hunk 6 above)
Comment (adrianna-chang-shopify) [reply challenge_13/action_dispatch.rb]: Ooh, the use of a `BodyProxy` is really interesting! Definitely on theme with our explorations into streaming responses - it's neat that the Common Logger middleware returns a body that is actual a callable BodyProxy, configured with a block, that can be invoked whenever the response is sent back to the client, at a much later point in time to our initial call to the Common Logger middleware!

File: challenge_13/action_dispatch.rb (hunk 8 above)
Comment (adrianna-chang-shopify) [reply challenge_13/action_dispatch.rb:33]: Right, this is handy to know - thanks!

File: challenge_13/action_dispatch.rb (hunk 5 above)
Comment (adrianna-chang-shopify) [reply challenge_13/action_dispatch.rb]: Let me give it a shot! 😉

File: challenge_13/config.ru [hunk 11]
```diff
@@ -20,7 +20,7 @@ class CharsetMiddleware
 
   def call(environment)
     response = Rack::Response[*@app.call(environment)]
-    response.content_type = "#{response.content_type}; charset=#{@charset}"
+    response.content_type = "#{response.content_type}; charset=#{@charset}" unless response.media_type_params.any?
```
Comment (tomstuart) [inline challenge_13/config.ru:23]: Great! Note that, in general, other media type parameters (e.g. [`boundary`](https://tools.ietf.org/html/rfc2046#section-5.1.1) in the case of `multipart/mixed`) might be present, so the _super_ correct thing here is to check for the presence of the `charset` parameter [specifically](https://github.com/adamluzsi/rack-charset/blob/v1.0.0/lib/rack/charset.rb#L16), but in practice other parameters are very rare so what you’ve done here will work fine.

Comment (adrianna-chang-shopify) [reply challenge_13/config.ru:23]: Ah, yep that makes sense! Thanks for making note of this ❤️
